Add the Asset Store component (#2430)
kyma-project/kyma: docs/asset-store/docs/06-01-asset.md

| **bucketref.name** |    **YES**   | Specifies the name of the bucket for storing the asset. |
| **spec.source.validationwebhookservice** |    **NO**   | Provides specification of the validation webhook service. |
| **spec.source.validationwebhookservice.name** |    **NO**   | Provides the name of the validation webhook service. |
| **spec.source.validationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
| **spec.source.validationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |
| **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the mutation webhook service. |
| **spec.source.mutationwebhookservice.name** |    **NO**   | Provides the name of the mutation webhook service. |
| **spec.source.mutationwebhookservice.namespace** |    **NO**   | Provides the Namespace in which the service is available. It must be the same as the asset's Namespace. |
| **spec.source.mutationwebhookservice.endpoint** |    **NO**   | Specifies the endpoint to which the service sends calls. |

> | **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed specification of the mutation service. |

Reviewer: ```suggestion
| **spec.source.mutationwebhookservice.metadata** |    **NO**   | Provides detailed metadata specific for a given mutation service and its functionality. |
```

---

The `assets.assetstore.kyma-project.io` CustomResourceDefinition (CRD) is a detailed description of the kind of data and the format used to define an asset to store in a cloud storage bucket. To get the up-to-date CRD and show the output in the `yaml` format, run this command:

```
kubectl get crd assets.assetstore.kyma-project.io -o yaml
```

## Sample custom resource

This is a sample resource (CR) that provides details of the bucket for storing assets.

```
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Asset
metadata:
  name: my-package-assets
  namespace: default
spec:
  source:
    mode: single
    url: https://some.domain.com/main.js
  bucketRef:
    name: my-bucket

```

### Validation and mutation webhook services

You can also define validation and mutation services:
- **Validation webhook** performs the validation of fetched assets before they are uploaded to the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status.

> - **Mutation webhook** acts similarly to the validation service. The difference is that it mutates the asset instead of just validating it. For example, this can be an asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.

Reviewer 2: ```suggestion
- **Mutation webhook** acts similarly to the validation service. The difference is that it mutates the asset instead of just validating it. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
```

---
title: Asset
type: Custom Resource
---

The `assets.assetstore.kyma-project.io` CustomResourceDefinition (CRD) is a detailed description of the kind of data and the format used to define an asset to store in a cloud storage bucket. To get the up-to-date CRD and show the output in the `yaml` format, run this command:

```
kubectl get crd assets.assetstore.kyma-project.io -o yaml
```

## Sample custom resource

This is a sample resource (CR) that provides details of the bucket for storing assets.

```
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Asset
metadata:
  name: my-package-assets
  namespace: default
spec:
  source:

>     mode: single / index / package

Author: ok, I only left single

```
apiVersion: assetstore.kyma-project.io/v1alpha1
kind: Asset
metadata:
  name: my-package-assets
  namespace: default
spec:
  source:
    mode: single | index | package
    url: https://some.domain.com/main.js
    validationWebhookService:
      name: swagger-operations-svc
      namespace: default
      endpoint: "/validate"
    mutationWebhookService:
      name: swagger-operations-svc
      namespace: default
      endpoint: "/mutate"
      metadata:
        rewrite: keyvalue
        pattern: \json|yaml
        data:
          basePath: /test/v2
  bucketRef:
    name: my-bucket
status:
  phase: Ready | Failed | Pending
  reason: ValidationFailure
  message: "The file is not valid against the provided json schema"
  lastHeartbeatTime: "2018-01-03T07:38:24Z"

```

## Custom resource parameters

This table lists all the possible parameters of a given resource together with their descriptions:

| Parameter   |      Mandatory      |  Description |
|:----------:|:-------------:|:------|

> | **metadata.name** |    **YES**   | Specifies the name of the CR. |

Author: yup I know these are basics but that's one of the parameters :(

> | **metadata.name** |    **YES**   | Specifies the name of the CR. |

Reviewer: ok, I think that sometimes we use the name of the resource for important things, like in Bucket CR, so people might sometimes want to know what is the name used for. I agree it must stay

| **metadata.namespace** |    **YES**   | Defines the Namespace in which the CR is available. |
| **spec.source.mode** |    **YES**   | Specifies if the asset is a single file, zip or tar, or a `index.yaml` file. For details on the **index** mode, see [this](#sample-index-file) section. |
| **spec.source.url** |    **YES**   | Specifies the location of the file. |
| **bucketref.name** |    **YES**   | Specifies the name of the bucket for storing the asset. |

> | **spec.source.validationwebhookservice** |    **NO**   | Provides specification of the validation webhook service. |

Reviewer: we are missing description of `spec.source.validationwebhookservice.name` and others like namespace, endpoint, metadata

| **spec.source.mutationwebhookservice** |    **NO**   | Provides specification of the validation webhook service. |

> 

Reviewer: we also need detailed description of status fields here

Reviewer: still not all status fields are described, like `assetRef`, `assetRef.assets` or `assetRef.baseUrl`

Author: ok, I've described them
